bmwqemu.pm

```diff
@@ -198,24 +197,24 @@ sub update_line_number {
 }
 
 # pretty print like Data::Dumper but without the "VAR1 = " prefix
-sub pp {
+sub pp(@args) {
     # FTR, I actually hate Data::Dumper.
-    my $value_with_trailing_newline = Data::Dumper->new(\@_)->Terse(1)->Useqq(1)->Dump();
+    my $value_with_trailing_newline = Data::Dumper->new(\@args)->Terse(1)->Useqq(1)->Dump();
```

Comment:
In cases like this (so this counts for `log_call` as well) signatures have zero benefit. I suppose they incur an additional copy of the argument list into an array. Considering those functions are called a lot (so the additional slow-down accumulates) it is likely not worth it.

Reply:
I wouldn't say _zero_ benefit.
You can see at one glance that the function takes an array, and this array is also easier to spot than the short `@_` in the body of the sub.
Currently signatures like that might indeed be a _bit_ slower because `@_` is still there, but `@_` will be dropped for performance reason in future versions of perl (only in signature subs of course).

Reply:
> Maybe, maybe not.

That's a bad justification for ignoring a possible performance regression. Considering the change is quite needless here I'd just avoid it.

(And yes, maybe I'm wrong and Perl optimizes this case and thus no copy is required. Annoyingly neither https://quick-bench.com nor https://godbolt.org are available for Perl.)

Reply:
https://metacpan.org/pod/Benchmark comes with perl.

I'm surprised that you are so worried about it. We (well, Oli mostly) added signatures to hundreds of subroutines by now.

Reply:
I'm not that worried. I just see that this change has a negative outcome considering the very little benefit signatures have here. Even if the negative outcome is very little it is still negative and thus I'm reluctant to think it is a good change. However, if you still want it, be welcome to dismiss my review.

Reply:
I would prefer if we can adopt the simple rule: "Signatures everywhere" which I would also like to protect with an according style check

Reply:
You must hate passing parameters in C++ then :-)
I suppose extending the rule by "except args are just passed through" doesn't make it much more complicated so not really an argument. And I suppose in case we use prototypes we already need to make an exception.

Reply:
I would be ok if you can
1. Convince everybody else
2. Provide an according style check in xt/01-style.t

Reply:
1. You could also try to convince me that there's no performance regression. "Maybe, maybe not." doesn't cut it.
2. So you're offloading the task to me?

But I don't care enough so I've just dismissed my review.

Reply:
```
#!/usr/bin/env perl
use v5.22;
use warnings;
use experimental qw/ signatures /;

use Data::Dumper;
use Benchmark qw/ timethese /;

sub old {
    my $value_with_trailing_newline = Data::Dumper->new(\@_)->Terse(1)->Useqq(1)->Dump();
}
sub signature (@x) {
    my $value_with_trailing_newline = Data::Dumper->new(\@x)->Terse(1)->Useqq(1)->Dump();
}

my @data = 'a' .. 'e'; # run 1
@data = "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb"; # run 2
@data = "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb" x 10; # run 3

timethese($ARGV[0] || -1, {
    old => sub { old(@data) for 1 .. ($ARGV[1] || 10) },
    signature => sub { signature(@data) for 1 .. ($ARGV[1] || 10) },
});
```
```
tina@linux-8vf8:/tmp/bench-signature% perl test.pl 300000 1 # run 1
Benchmark: timing 300000 iterations of old, signature...
       old:  2 wallclock secs ( 2.28 usr +  0.00 sys =  2.28 CPU) @ 131578.95/s (n=300000)
 signature:  3 wallclock secs ( 2.43 usr +  0.00 sys =  2.43 CPU) @ 123456.79/s (n=300000)
tina@linux-8vf8:/tmp/bench-signature% perl test.pl 300000 1 # run 2
Benchmark: timing 300000 iterations of old, signature...
       old:  2 wallclock secs ( 1.80 usr +  0.00 sys =  1.80 CPU) @ 166666.67/s (n=300000)
 signature:  2 wallclock secs ( 1.84 usr +  0.00 sys =  1.84 CPU) @ 163043.48/s (n=300000)
tina@linux-8vf8:/tmp/bench-signature% perl test.pl 300000 1 # run 3
Benchmark: timing 300000 iterations of old, signature...
       old:  3 wallclock secs ( 2.37 usr +  0.00 sys =  2.37 CPU) @ 126582.28/s (n=300000)
 signature:  2 wallclock secs ( 2.41 usr +  0.00 sys =  2.41 CPU) @ 124481.33/s (n=300000)
```

Reply:
I ran the benchmark as well but with different modules to resemble better what we use:

```
use strictures;
use Mojo::Base -strict, -signatures;
```

this gives for 500k iterations 99s (old) vs. 95s (signatures) so signatures are slightly faster even. Can we agree on using signatures everywhere? If not I can also leave out the corresponding changes of the two functions where I used `@args` and we can discuss that separately.

Reply:
it sounds unlikely to me that signatures are faster.
can you post your benchmark? i posted mine and signatures were slightly slower (but not enough that i would argue against them).

the usage of `strictures` or the way we enable signatures should not have any effect on the result.

Reply:
Yes, `strictures` shouldn't matter. Full results for the three test arrays 1, 2, 3 in this order:

```
Benchmark: timing 500000 iterations of old, signature...
       old: 82 wallclock secs (82.37 usr +  0.00 sys = 82.37 CPU) @ 6070.17/s (n=500000)
 signature: 89 wallclock secs (89.12 usr +  0.01 sys = 89.13 CPU) @ 5609.78/s (n=500000)

Press ENTER or type command to continue
Benchmark: timing 500000 iterations of old, signature...
       old: 97 wallclock secs (96.75 usr +  0.12 sys = 96.87 CPU) @ 5161.56/s (n=500000)
 signature: 101 wallclock secs (100.41 usr +  0.00 sys = 100.41 CPU) @ 4979.58/s (n=500000)

Press ENTER or type command to continue
Benchmark: timing 500000 iterations of old, signature...
       old: 68 wallclock secs (67.82 usr +  0.00 sys = 67.82 CPU) @ 7372.46/s (n=500000)
 signature: 67 wallclock secs (67.36 usr +  0.01 sys = 67.37 CPU) @ 7421.70/s (n=500000)
```

so signatures seem to be slower for the first and second set but faster for the third set although I would say that all cases are not significant.